modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/CacheGroupPageScanner.java

```diff
@@ -0,0 +1,479 @@
+            forEachPageStore(grp, new IgniteInClosureX<Integer>() {
+                @Override public void applyx(Integer partId) throws IgniteCheckedException {
+                    int pagesCnt = ctx.cache().context().pageStore().pages(grp.groupId(), partId);
+
+                    partStates[Math.min(partId, partStates.length - 1)] = pagesCnt;
+                }
+            });
+        } finally {
+            ctx.cache().context().database().checkpointReadUnlock();
+        }
+
+        return partStates;
+    }
+
+    /**
+     * @param grp Cache group.
+     * @param hnd Partition handler.
+     */
+    private void forEachPageStore(CacheGroupContext grp, IgniteInClosureX<Integer> hnd) throws IgniteCheckedException {
+        int parts = grp.affinity().partitions();
+
+        IgnitePageStoreManager pageStoreMgr = ctx.cache().context().pageStore();
+
+        for (int p = 0; p < parts; p++) {
+            if (!pageStoreMgr.exists(grp.groupId(), p))
+                continue;
+
+            hnd.applyx(p);
+        }
+
+        hnd.applyx(PageIdAllocator.INDEX_PARTITION);
+    }
+
+    /**
+     * Cache group partition scanning task.
+     */
+    private class GroupScanTask extends GridFutureAdapter<Void> implements Runnable {
+        /** Cache group ID. */
+        private final CacheGroupContext grp;
+
+        /** Partition IDs. */
+        private final Set<Integer> parts;
+
+        /** Page memory. */
+        private final PageMemoryEx pageMem;
+
+        /**
+         * @param grp Cache group.
+         */
+        public GroupScanTask(CacheGroupContext grp, Set<Integer> parts) {
+            this.grp = grp;
+            this.parts = new GridConcurrentHashSet<>(parts);
+
+            pageMem = (PageMemoryEx)grp.dataRegion().pageMemory();
+        }
+
+        /** {@inheritDoc} */
+        @Override public synchronized boolean cancel() throws IgniteCheckedException {
+            return onCancelled();
+        }
+
+        /**
+         * Stop reencryption of the specified partition.
+         *
+         * @param partId Partition ID.
+         * @return {@code True} if reencryption was cancelled.
+         */
+        public synchronized boolean excludePartition(int partId) {
+            return parts.remove(partId);
+        }
+
+        /**
+         * @return Cache group ID.
+         */
+        public int groupId() {
+            return grp.groupId();
+        }
+
+        /** {@inheritDoc} */
+        @Override public void run() {
+            try {
+                for (int partId : parts) {
+                    long state = ctx.encryption().getEncryptionState(grp.groupId(), partId);
+
+                    if (state == 0)
+                        continue;
+
+                    scanPartition(partId, ReencryptStateUtils.pageIndex(state), ReencryptStateUtils.pageCount(state));
+
+                    if (isDone())
+                        return;
+                }
+
+                boolean added = cpWaitGrps.add(this);
```

Comment:
cpWaitGrps is a collection of groups waiting for a checkpoint to start.

PageScanner performs the following steps:
1. mark all pages in the group as dirty
2. wait for the start of the checkpoint (onCheckpointBegin)
3. wait for finishedStateFut() of the current checkpoint to complete

I though that after finishedStateFut finishes, all pages (dirty at the time of starting the checkpoint) are written to disk, am I missing something?

Reply:
I see that we cannot guarantee which pages included in that checkpoint. Because onCheckpointBegin method is executed out of write lock, that means some dirty pages will belong to another checkpoint (next after that).

Reply:
@vldpyatkov thanks, seems that we should use beforeCheckpointBegin instead of onCheckpointBegin, right?

Reply:
Now, we can get information about a written group after than that happens.
If it is ok, that is right. I still badly oriented in these patch.